### PR TITLE
Implemented ComplexPlane Class: Complex Sets

### PR DESCRIPTION
**Pull request by aktech**
Initial version of ComplexPlane Class

@flacjacket @hargup 

TODO:
- [x] ComplexPlane definition
- [x] Add rules for Intersection & Union with Real Sets
- [x] `normalize_theta_set`
- [x] Add Tests
- [x] Add Printer
- [x] **100%** [Coverage](https://github.com/sympy/sympy/pull/9463#issuecomment-113172770)

Other TODO: (Maybe in an another PR)
- Special Printer for ComplexPlane (with custom Intervals)

### CODE REVIEW

**Comment by hargup** [comment]
+1 for the doc-strings.

It will be better if you specify input in polar form by keyword arguments like `polar=True` rather than the type of input.

**Comment by aktech** [comment]
> It will be better if you specify input in polar form by keyword arguments like polar=True rather than the type of input.

Done.

File: sympy/core/tests/test_args.py [hunk 1]
```diff
@@ -468,8 +468,8 @@ def test_sympy__sets__sets__Complement():
 
 def test_sympy__sets__sets__SymmetricDifference():
     from sympy.sets.sets import FiniteSet, SymmetricDifference
-    assert _test_args(SymmetricDifference(FiniteSet(1, 2, 3), \
-           FiniteSet(2, 3, 4)))
+    assert _test_args(SymmetricDifference(FiniteSet(1, 2, 3),
```
Comment (hargup) [inline sympy/core/tests/test_args.py]: Please avoid pure formatting changes, they mess up `git-blame`.

Comment (aktech) [inline sympy/core/tests/test_args.py]: It wasn't pure formatting though, I have added `args` tests for  ComplexPlane Class, in that file.
I read somewhere in mailing list discussion by **jcrist** , It's good to make formatting changes around the code we are playing with. Isn't it?

Comment (toolforger) [inline sympy/core/tests/test_args.py]: I agree that formatting can be done if the code is changed anyway; however, in this case
1) You aren't changing the code near these lines (though you do remove a superfluous backslash),
2) the new formatting is not an actual improvement (both ways to format the code are PEP8-compliant, but the changed formatting style is sometimes considered less good).

Strictly speaking for myself and not for the project, I'd recommend keeping subexpressions on one line as far as possible, and indent them by multiples of 4. Something like this:

```
assert _test_args(
    SymmetricDifference(FiniteSet(1, 2, 3), FiniteSet(2, 3, 4)))
```

Still, this isn't really a change related to your other work; I wouldn't -1 the PR if it is in, but I'd recommend keeping formatting away from "real work" so that people don't need to scroll over it when they're looking for what's in the PR description.

Comment (aktech) [inline sympy/core/tests/test_args.py]: Thanks, removed formatting!

File: sympy/sets/fancysets.py [hunk 2]
```diff
@@ -443,3 +444,126 @@ def _sup(self):
...
+            from sympy import cos, sin
+            return ImageSet.__new__(cls, Lambda((r, theta),
+                                    r*(cos(theta) + I*sin(theta))), a*b)
+
+    def _contains(self, other):
+        lambda_arg = self.args[0].args[1]
+        a, b = self.args[1].args
+
+        # self in rectangular form
+        if lambda_arg.is_Add:
```
Comment (hargup) [inline sympy/sets/fancysets.py]: Use the more descriptive `self.polar` to check the form of the representation.

File: sympy/sets/fancysets.py [hunk 3]
```diff
@@ -443,3 +444,126 @@ def _sup(self):
...
+        if polar is False:
+            return ImageSet.__new__(cls, Lambda((x, y), x + I*y), a*b)
+
+        # Polar Form
+        elif polar is True:
+            from sympy import cos, sin
+            return ImageSet.__new__(cls, Lambda((r, theta),
+                                    r*(cos(theta) + I*sin(theta))), a*b)
+
+    def _contains(self, other):
```
Comment (hargup) [inline sympy/sets/fancysets.py]: As @moorepants suggested in https://github.com/sympy/sympy/pull/9438#discussion_r31336944, it is recommended that you write the tests before you write the code for a particular feature, it will the expectations from the code clear and it will also make it easier for us to test if your code is correct or not.

I saw that you have written doctests which partially solves the purpose but unit tests gives a clearer picture of the  behaviour of code.

File: sympy/sets/fancysets.py [hunk 4]
```diff
@@ -443,3 +444,126 @@ def _sup(self):
     @property
     def _boundary(self):
         return self
+
+
+def as_r_theta(z):
```
Comment (hargup) [inline sympy/sets/fancysets.py]: We don't need this function, we already have `arg` [1] and `Abs` [2] as a part of elementary functions.

[1]: http://docs.sympy.org/dev/modules/functions/elementary.html#arg
[2]: http://docs.sympy.org/dev/modules/functions/elementary.html#abs

File: sympy/sets/fancysets.py (hunk 3 above)
Comment (aktech) [inline sympy/sets/fancysets.py]: Yes, I will keep that in mind from now.

File: sympy/sets/fancysets.py (hunk 4 above)
Comment (aktech) [inline sympy/sets/fancysets.py]: Yes, thanks!

File: sympy/sets/fancysets.py (hunk 2 above)
Comment (aktech) [inline sympy/sets/fancysets.py]: Sure.

File: sympy/sets/fancysets.py (hunk 3 above)
Comment (aktech) [inline sympy/sets/fancysets.py]: I have added some tests now.

File: sympy/sets/tests/test_fancysets.py [hunk 5]
```diff
@@ -258,3 +258,61 @@ def test_ImageSet_simplification():
...
+    right_half_plane = ComplexPlane(Interval(0, oo), Interval(-oo, oo))
+    right_quart_plane = ComplexPlane(Interval(0, oo), Interval(0, oo))
+
+    assert upper_half_plane.intersect(unit_square) == upper_half_unit_square
+    assert right_half_plane.intersect(right_quart_plane) == right_quart_plane
+    assert upper_half_plane.intersect(right_half_plane) == right_quart_plane
+    assert upper_half_plane.intersect(lower_half_plane) == EmptyPlane
+
+    # Intersection Polar and Rectangular
+    raises(ValueError, lambda: unit_disk.intersect(unit_square))
```
Comment (hargup) [inline sympy/sets/tests/test_fancysets.py]: It should return an unevaluated `Intersection` object rather raising an error.

**Comment by leosartaj** [comment]
It would be good if user could do something like

```
>>> ComplexPlane((2, 3), (4, 6)) 
```

Rather than

```
>>> a = Interval(2, 3)
>>> b = Interval(4, 6)
>>> ComplexPlane(a, b)
```
Maybe args can be handled as ``Interval``.
I would like to not import ``Interval``, everytime I have to make an instance of ``ComplexPlane``.

**Comment by aktech** [comment]
@leosartaj 
With the current API, your suggestions could have been incorporated, but the current User Interface is not good, as it will not suffice in representing all possible regions in a Complex Plane, So I am in a process of changing the input API, which is ProductSets or Union of ProductSets and it's probably not good to simplify UI in that case.

**Comment by aktech** [comment]
@hargup  @flacjacket  Please have a look.

File: sympy/sets/fancysets.py [hunk 6]
```diff
@@ -443,3 +445,309 @@ def _sup(self):
...
+        [2, 3] x [4, 5]
+        >>> C2 = ComplexPlane(Union(a*b, b*c))
+        >>> C2.sets
+        [2, 3] x [4, 5] U [4, 5] x [1, 7]
+
+        """
+        return self.args[1]
+
+    @property
+    def psets(self):
```
Comment (hargup) [inline sympy/sets/fancysets.py]: The `sets` function should be enough, it is trivial to convert between `psets` and `sets`.

Comment (aktech) [inline sympy/sets/fancysets.py]: I was expecting this, but it saves writing same lines of code again and again further.
`psets` gives usable form of the sets.

Comment (hargup) [inline sympy/sets/fancysets.py]: Same for `a_interval` and `b_interval`, they are not really needed and you should only one function that gives the input set.

Comment (aktech) [inline sympy/sets/fancysets.py]: Same as this [comment](https://github.com/sympy/sympy/pull/9463/files#r32110887).
Increases readability and simplifies code further in other methods.

File: sympy/sets/fancysets.py [hunk 7]
```diff
@@ -443,3 +445,309 @@ def _sup(self):
...
+        """
+        b_interval = []
+        for element in self.psets:
+            b_interval.append(element.args[1])
+
+        b_interval = Union(*b_interval)
+        return b_interval
+
+    @property
+    def polar(self):
```
Comment (hargup) [inline sympy/sets/fancysets.py]: You can also assign `self.polar` to the `polar` argument in the `__new__` method, also the calculation is wrong because as both polar and cartesian form takes the input as Product sets.

Comment (aktech) [inline sympy/sets/fancysets.py]: > You can also assign self.polar to the polar argument in the __new__ method,

That would not work, as that creates a class variable & it gets overridden every-time a new instance is called.
That's why I have used an OOP style property.

> also the calculation is wrong because as both polar and cartesian form takes the input as Product sets.

No, Since the `self` here is the returned object, the `arg` is as follows:
 
`self.args[0].args[1]` is  `x + I*y`  (Add Object) for rectangular.

`self.args[0].args[1]` is `r*(I*sin(theta) + cos(theta))`  (Mul object) for polar.

File: sympy/sets/fancysets.py [hunk 8]
```diff
@@ -443,3 +445,309 @@ def _sup(self):
...
+                return ComplexPlane(new_r_interval*new_theta_interval,
+                                    polar=True)
+
+        if other.is_Interval:
+            new_interval = []
+
+            # self in rectangular form
+            if self.polar is False:
+                for element in self.psets:
+                    if (S.Zero, S.Zero) in element:
```
Comment (hargup) [inline sympy/sets/fancysets.py]: You only need to check if the imaginary part is zero, real part need not be zero.

File: sympy/sets/fancysets.py [hunk 9]
```diff
@@ -443,3 +445,309 @@ def _sup(self):
...
+
+        # Rectangular Form
+        if polar is False:
+            obj = ImageSet.__new__(cls, Lambda((x, y), x + I*y),
+                                   sets)
+
+        # Polar Form
+        elif polar is True:
+            from sympy import cos, sin
+            obj = ImageSet.__new__(cls, Lambda((r, theta),
```
Comment (hargup) [inline sympy/sets/fancysets.py]: Theta should be normalized to be in set `[0, pi)`.

File: sympy/sets/fancysets.py [hunk 10]
```diff
@@ -443,3 +445,309 @@ def _sup(self):
...
+                for element in self.psets:
+                    if (S.Zero, S.Zero) in element:
+                        new_interval.append(element.args[0])
+                new_interval = Union(*new_interval)
+                return Intersection(new_interval, other)
+
+            # self in polar form
+            elif self.polar:
+                for element in self.psets:
+                    if (0 in element.args[1]) or (S.Pi in element.args[1]):
```
Comment (hargup) [inline sympy/sets/fancysets.py]: You need not check for Pi if theta is normalized. If it is not normalized then you will have to check for all `n*Pi` where `n` is any integer.

File: sympy/sets/fancysets.py (hunk 9 above)
Comment (aktech) [inline sympy/sets/fancysets.py]: Sorry, I have not documented that, but I am expecting theta to be in `[0, 2*Pi)` .

File: sympy/sets/fancysets.py (hunk 10 above)
Comment (aktech) [inline sympy/sets/fancysets.py]: See [this](https://github.com/sympy/sympy/pull/9463#discussion_r32112842)

File: sympy/sets/fancysets.py [hunk 11]
```diff
@@ -443,3 +445,309 @@ def _sup(self):
...
+                r2, theta2 = other.a_interval, other.b_interval
+                new_r_interval = Intersection(r1, r2)
+                new_theta_interval = Intersection(theta1, theta2)
+                # 0 and 2*Pi means the same
+                if (2*S.Pi in theta1 and S(0) in theta2) or (2*S.Pi in theta2 and S(0) in theta1):
+                    new_theta_interval = Union(new_theta_interval, FiniteSet(0))
+                return ComplexPlane(new_r_interval*new_theta_interval,
+                                    polar=True)
+
+        if other.is_Interval:
```
Comment (hargup) [inline sympy/sets/fancysets.py]: So the general idea is to take real values in the complex set and take their intersection with the real set provided. In that case you real set need not be an Interval, it can be any general real set.

Comment (aktech) [inline sympy/sets/fancysets.py]: What are the other forms expected?
`Union` ?

File: sympy/sets/fancysets.py (hunk 8 above)
Comment (aktech) [inline sympy/sets/fancysets.py]: Imagine, a rectangle in a Complex plane:
It can only contain Real points in it, if and only if origin is inside that rectangle.

So, (0, 0) must be present in the `ComplexPlane`.

Comment (aktech) [inline sympy/sets/fancysets.py]: Yes, thanks!

File: sympy/sets/fancysets.py (hunk 9 above)
Comment (hargup) [inline sympy/sets/fancysets.py]: Sorry my bad, it is `[0, 2*Pi)`

File: sympy/sets/fancysets.py (hunk 11 above)
Comment (hargup) [inline sympy/sets/fancysets.py]: It can be anything `Union`, `Intersection` or `ImageSet`, You need not check for the type of `other` you can simply do `other in S.Reals`.

Comment (aktech) [inline sympy/sets/fancysets.py]: OK, thanks!

File: sympy/sets/fancysets.py (hunk 9 above)
Comment (aktech) [inline sympy/sets/fancysets.py]: I think we should raise a `ValueError` , when `theta` is not in `[0, 2*Pi)` , with a message " *Please enter theta in the range* `[0, 2*Pi)`  "
So, that we can avoid pre-processing of theta & avoid confusion as well.
thoughts?

Comment (hargup) [inline sympy/sets/fancysets.py]: Nope `ValueError` won't be correct as `theta` outside `[0, 2*pi)` are also valid values. You need to express theta in form of `2*n*pi + x` where `n` is an integer and `x` is a value in `[0, 2*pi)` and use that `x`. As far as I remember there is a method in the `Trigonometry` module which does this normalization thing.

Comment (aktech) [inline sympy/sets/fancysets.py]: Got it, `_pi_coeff` does that.
Thanks!

**Comment by aktech** [comment]
Changed `normalize_theta_Interval` to `normalize_theta_set`.
It now accepts FiniteSet & Interval.

File: sympy/sets/fancysets.py [hunk 12]
```diff
@@ -443,3 +445,405 @@ def _sup(self):
...
+                    new_theta.append(S.Zero)
+            else:
+                new_theta.append(k*S.Pi)
+        return FiniteSet(*new_theta)
+
+    elif theta.is_subset(S.Reals):
+        raise NotImplementedError("Normalizing theta when, its %s is not"
+                                  "Implemented" % type(theta))
+    else:
+        raise ValueError(" %s does not a belongs to S.Reals" % (theta))
```
Comment (hargup) [inline sympy/sets/fancysets.py]: Better say `%s is not a real set`, none of the sets, including Intervals, belongs to S.Reals as its members are only numbers.

File: sympy/sets/tests/test_fancysets.py [hunk 13]
```diff
@@ -258,3 +259,119 @@ def test_ImageSet_simplification():
...
+    c8 = ComplexPlane(Interval(12, 16)*Interval(14, 20))
+
+    p3 = Union(Interval(2, 5)*Interval(6, 9), Interval(4, 6)*Interval(10, 12))
+    p4 = Union(Interval(0, 10)*Interval(-10, 0), Interval(12, 16)*Interval(14, 20))
+
+    assert c5.union(c6) == ComplexPlane(p3)
+    assert c7.union(c8) == ComplexPlane(p4)
+
+
+def test_normalize_theta_set():
```
Comment (hargup) [inline sympy/sets/tests/test_fancysets.py]: Also add a test to check if it raises a `ValueError` for non real sets.

**Comment by hargup** [comment]
`normalize_theta_set` is fine. Can you include a coverage report for your code?

**Comment by aktech** [comment]
coverage-complex-sets

**Comment by aktech** [comment]
>  Can you include a coverage report for your code?

It decreased by 1% , from master should I add more tests?

**Comment by toolforger** [comment]
You should cover 100% of your new code.
Minus comments, and maybe (some people disagree) minus trivial lines such as `week = 7 * day`.

**Comment by aktech** [comment]
### Coverage of new Code:  94%
 See  `sympy/sets/complexplane.py`

Only 8 lines are missing, I assume they are trivial(?).

screenshot from 2015-06-13 20 34 08

**Comment by aktech** [comment]
@hargup 
The issue with cyclic import while creating `S.Complex` , is resolved as [this](https://github.com/toolforger/sympy/commit/4264f649f39b3d20ea0f1b4193171c4cec62435c) is Merged.

**Comment by toolforger** [comment]
> The issue with cyclic import while creating `S.Complex` , is resolved as [this](https://github.com/toolforger/sympy/commit/4264f649f39b3d20ea0f1b4193171c4cec62435c) is Merged.

Heh. I didn't expect my work to bear fruit before it's even merged :-)

Incidentally, it just got merged. You should be able to see it if you 
rebase.

**Comment by aktech** [comment]
@hargup  @flacjacket Please have a look.

File: sympy/sets/fancysets.py [hunk 14]
```diff
@@ -443,3 +445,451 @@ def _sup(self):
...
+                if val == S.Zero:
+                    new_theta.append(S.Zero)
+                else:
+                    # when theta is n*pi
+                    new_theta.append(2*S.Pi)
+            else:
+                new_theta.append(k*S.Pi)
+        # one complete circle
+        if (new_theta[0] == new_theta[1]):
+            return Interval(0, 2*S.Pi, False, True)
```
Comment (hargup) [inline sympy/sets/fancysets.py]: This line is not covered.

File: sympy/sets/fancysets.py [hunk 15]
```diff
@@ -443,3 +445,451 @@ def _sup(self):
...
+        >>> C2 = ComplexPlane(Union(a*b, b*c))
+        >>> C2.psets
+        ([2, 3] x [4, 5], [4, 5] x [1, 7])
+
+        """
+        if self.args[1].is_ProductSet:
+            psets = ()
+            psets = psets + (self.args[1], )
+        else:
+            psets = self.args[1].args
```
Comment (hargup) [inline sympy/sets/fancysets.py]: This line is not covered.

**Comment by hargup** [comment]
@aktech run `./bin/coverage_report.py sympy/sets` to generate the html report, if click on the module it will show the lines not covered in red. Except 4 lines, your code is fully covered by your test suite, which is cool thing. Adding tests for those lines should be trivial. After that, I'm +1 to for a merge.

File: sympy/sets/fancysets.py [hunk 16]
```diff
@@ -443,3 +445,451 @@ def _sup(self):
...
+                return Intersection(new_interval, other)
+
+            # self in polar form
+            elif self.polar:
+                for element in self.psets:
+                    if (0 in element.args[1]) or (S.Pi in element.args[1]):
+                        new_interval.append(element.args[0])
+                new_interval = Union(*new_interval)
+                return Intersection(new_interval, other)
+            return None
```
Comment (hargup) [inline sympy/sets/fancysets.py]: This and [this](https://github.com/sympy/sympy/pull/9463/files#diff-c1c17bdd28aafd400bf9415c7d048382R888) is also not covered. Tests for these should check if we are getting an unevaluated `Union` or `Intersection` object if the evaluation fails.

File: sympy/sets/fancysets.py (hunk 14 above)
Comment (aktech) [inline sympy/sets/fancysets.py]: I just noticed, these two lines aren't needed anymore, since we already covered the case of complete circle above in [Line 493](https://github.com/sympy/sympy/pull/9463/files#diff-c1c17bdd28aafd400bf9415c7d048382R493). I would remove these.

File: sympy/sets/fancysets.py (hunk 16 above)
Comment (aktech) [inline sympy/sets/fancysets.py]: I am adding tests for unevaluated `Union` or `Intersection` object if the evaluation fails, but we don't need `return None`  here for `_intersect` method, as unevaluated object is already returned by the call to `Intersection` class.

**Comment by aktech** [comment]
## Coverage of ComplexPlane: 100%

complexplane-coverage

**Comment by aktech** [comment]
Test Timeouts:
* https://travis-ci.org/sympy/sympy/jobs/67375133
* https://travis-ci.org/sympy/sympy/jobs/67375131

File: sympy/sets/tests/test_fancysets.py [hunk 17]
```diff
@@ -177,7 +179,7 @@ def test_fun():
         Range(-10, 11))) == FiniteSet(-1, -sqrt(2)/2, 0, sqrt(2)/2, 1))
 
 
-def test_reals():
+def test_Reals():
```
Comment (hargup) [inline sympy/sets/tests/test_fancysets.py]: This is unrelated to this PR, please avoid such changes from the next time.

**Comment by hargup** [comment]
@aktech Good work :+1: , it is in now.

**Comment by aktech** [comment]
Thanks @hargup !
